wicker/core/storage.py

```diff
@@ -463,92 +487,141 @@ class S3PathFactory(WickerPathFactory):
                     - part-1-attempt-2345.parquet
     """
 
-    def __init__(self, s3_root_path: Optional[str] = None) -> None:
+    def __init__(self, s3_root_path: Optional[str] = None, prefix_replace_path: str = "") -> None:
         """Init S3PathFactory.
 
         Args:
             s3_root_path (Optional[str], optional): path to s3 root path. Defaults to None.
+            prefix_replace_path (str, optional): path to replace for given prefix. In this class this is generally
+                replacing s3 when a bucket is mounted to instance. Defaults to "".
         """
         s3_config = get_config().aws_s3_config
         store_concatenated_bytes_files_in_dataset = s3_config.store_concatenated_bytes_files_in_dataset
         s3_root_path = s3_root_path if s3_root_path is not None else s3_config.s3_datasets_path
         # ignore type as we already handled none case above
-        super().__init__(s3_root_path, store_concatenated_bytes_files_in_dataset)  # type: ignore
+        super().__init__(
+            root_path=s3_root_path,
+            prefix_replace_path=prefix_replace_path,
+            store_concatenated_bytes_files_in_dataset=store_concatenated_bytes_files_in_dataset,
+        )  # type: ignore
 
-    def get_dataset_assets_path(self, dataset_id: DatasetID, s3_prefix: bool = True) -> str:
+    @staticmethod
+    def __determine_cut_prefix(s3_prefix: bool, cut_prefix_override: Optional[str]):
+        """Determine the cut prefix by using default logic and if override provided preferring.
+
+        Args:
+            s3_prefix (bool, optional): Whehter to keep the s3 prefix or not.
+            cut_prefix_override (str, optional): Optional arbitrary prefix to prefer over s3 prefix.
+                used if provided.
+
+        Returns:
+            str: Prefix to trim off front of path.
+        """
+        # keep original logic
+        prefix_to_trim = "s3://" if not s3_prefix else None
```

Comment:
it looks to me that what we we really want to do is to somehow extend how `s3_prefix` operates, i.e, `(s3_prefix, cut_prefix_override)` is really one parameter, but we are doing via a series of patched overrides. Could you expand on the technical motivation for this approach?

Reply:
No super strong opinion here, but if someone wants to grab this change they should change their code anyways to provide `cut_prefix_override` right? In that case why not just expecting them to change their code to use a different class name that does this explicitly?  i.e, since `S3PathFactory` is small, why not inheriting directly from `WickerPathFactory` to provide this functionality? (not necessarily saying we should, but just like to understand better)

It looks like we are expecting `S3PathFactory` to support a lot more than `S3` :)

Reply:
The reasoning here is that your prefix to trim might not always be s3:// even in the case of using AWS or S3. Say you mount a bucket with `s3mountpoint` at a file location. That bucket path has now become `<file-system-path>` and you need to replace that entire prefix. That is the use case. 

They don't have to use the cut_prefix_override, it's only there to allow for mount point usage if the user specifies. It's not included in the WickerPathFactory because this overriding of the default is only needed in S3PathFactory because we have to backward compatibly make the change. The override is only there for that, nothing more.